### ISSUE TITLE
model size keeps the same after pruning or scaling in custom ai

### ISSUE THREAD

**Issue by piuz** [CoCoPIE-Group/XGen-Report#17]
I use a resnet18 to test the custom ai on cifar-10. 
[scripts.zip](https://github.com/CoCoPIE-Group/XGen-Report/files/9629205/scripts.zip)

Here are some results. 
![image](https://user-images.githubusercontent.com/75514414/191844748-0fcff8a3-98b2-465c-96f6-17b54091af55.png)

I found that the model size keeps the same after pruning or scaling, even though the inference speed might be different. Maybe you can have a check whether this is a bug.

**Comment by xipengshen** [comment]
@piuz Please post the log files.

**Comment by piuz** [comment]
[xgen_log_prune_cifar.txt](https://github.com/CoCoPIE-Group/XGen-Report/files/9637510/xgen_log_prune_cifar.txt)
[xgen_log_scale_cifar.txt](https://github.com/CoCoPIE-Group/XGen-Report/files/9637511/xgen_log_scale_cifar.txt)

Here are the log files for prune and scale on cifar-10 with resnet18.

**Comment by xipengshen** [comment]
@uniyushu @gdingcoco Please take a look.

**Comment by xipengshen** [comment]
Besides the size of the model, the substantial accuracy drops look strange.

**Comment by xipengshen** [comment]
@piuz It seems that you set the desired latency to 10ms. It explains why XGen won't prune or scale much as the original model already has a latency close to that. You may try an even smaller latency (e.g., 5ms) to see what happens. 
In addition, I saw that the accuracy on MNIST from XGen is pretty good. The low accuracy on cifar10 might be caused by the insufficient number of epochs? Can you try to increase the number of epoches?

**Comment by piuz** [comment]
OK. I will do more test to reduce the latency.

**Comment by piuz** [comment]
[XGen test - customai.pdf](https://github.com/CoCoPIE-Group/XGen-Report/files/9641967/XGen.test.-.customai.pdf)

I got some new results with target latency set to 5ms and 3ms as attached. There are still some issues. 

1) The model size still keeps the same. But it should be smaller. 

2) The accuracy is still low. I am not sure why the accuracy is so low. The dense model can reach high accuracy like 80% within 20 epochs. 

3) Pruning seems to reduce the latency successfully to meet the target latency. But scale does not seem to be able to reduce the latency. The latency of scaling seems to keep the same as that of  the dense model.

I attached the log files. 

[xgen_log_prune_5ms.txt](https://github.com/CoCoPIE-Group/XGen-Report/files/9641971/xgen_log_prune_5ms.txt)
[xgen_log_scale_5ms.txt](https://github.com/CoCoPIE-Group/XGen-Report/files/9641972/xgen_log_scale_5ms.txt)

**Comment by xipengshen** [comment]
@gdingcoco please take a look.

**Comment by gdingcoco** [comment]
It looks like pruning work well, and the reson of the model size not change is beacase it means "number of total parameters"  not  "number of total none zero parameters ". 

For scaling, please provide the database.json like last time

**Comment by piuz** [comment]
[database.zip](https://github.com/CoCoPIE-Group/XGen-Report/files/9648155/database.zip)

The database.json for scaling is attached.

**Comment by piuz** [comment]
I can achieve high accuracy on CIFAR with proper hyperparameters. The accuracy issue can be solved.

**Comment by xipengshen** [comment]
@piuz Did you try lower latencies (e.g., 5ms) on CIFAR? Can you see speedups and reasonable accuracies?

**Comment by piuz** [comment]
![333](https://user-images.githubusercontent.com/75514414/192619124-b4ef14a8-1163-4c6e-85a9-cb97ae80d6bf.jpg)

Yeah, I tried lower target latencies. For pruning, the latency can be reduced and the accuracy can be trained to reach high accuracy.  

For the scaling, I run the default 'III resnet (CIFAR)'. If the target latency is set to 3ms, the results does not seem to reduce the latency. It is still 12ms close to the original 12ms latency. I attached my configuration in the following. I am not asked to input or modify any hyperparameters such as scaling_factor. I just need to make some simple choice.  Maybe I did not use a proper configuration for scaling. I will do more check. 
![444](https://user-images.githubusercontent.com/75514414/192620252-069b5f51-e6e3-4a08-aaae-1be1ba61ff40.jpg)

**Comment by xipengshen** [comment]
@xipengshen Make scaling instructions in the document more clear.

**Comment by xipengshen** [comment]
@piuz We'd like to use the MNIST model as a quick-start example of XGen. Can you share the code (the version after you made it compatible with XGen) and dataset? We can name the model 'DigitNet' as it is for recognizing digits. Please try to reduce the number of epochs in the configuration file to a small number such that the execution can finish in just several minutes. My impression is that the model achieves very good accuracy even after just one epoch. 

@xinzhang-cocopie Can you work with Pu to add the model into our XGen Common AI model list? In the model list, we can name the model 'Image Classification 0: DigitNet (MNIST)'.

**Comment by piuz** [comment]
Thanks. I attach the scripts for MNIST including train script, pretrained model and xgen.json below. It works with an old version of XGen.
  [xgen_mnist.zip](https://github.com/CoCoPIE-Group/XGen-Report/files/9693404/xgen_mnist.zip)

Now I upgrade xgen to the newest version. A new problem shows up. For the customai with MNIST, I can not run compatibility test, pruning or scaling. I attached the log files. Maybe something are not properly configured. 

[20221002194936.zip](https://github.com/CoCoPIE-Group/XGen-Report/files/9693430/20221002194936.zip)
[20221002195732.zip](https://github.com/CoCoPIE-Group/XGen-Report/files/9693431/20221002195732.zip)

**Comment by xipengshen** [comment]
@piuz Thanks! Can you also share the data folder?

**Comment by piuz** [comment]
data is just mnist, which can be downloaded within one minute. The program will download the data automatically in first time running.

**Comment by xipengshen** [comment]
Got it. Thanks.

**Comment by xipengshen** [comment]
@piuz This might be also due to the expired compiler problem. After upgrading XGen, can you try this again?

**Comment by xipengshen** [comment]
@piuz I found out the reason for the problem. Our XGen updated its standard for the changes to the training script. The script you have is out of date. The updated instructions are [here](https://cocopieai-my.sharepoint.com/:b:/r/personal/xipeng_shen_cocopie_ai/Documents/XGen_4_Usage.pdf?csf=1&web=1&e=hwfa15).

**Comment by xipengshen** [comment]
@piuz I played with the mnist_cnn model a bit. The model training script has some problems. I made some changes and now have it working in all modes (compatibility_test, pruning, scaling). The fixed model files are [here](https://cocopieai-my.sharepoint.com/:u:/r/personal/xipeng_shen_cocopie_ai/Documents/XGen/mnist_cnn.tar?csf=1&web=1&e=1G98p5).
